pkg/daemon/ceph/osd/remove.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// RemoveOSDs purges a list of OSDs from the cluster
+func RemoveOSDs(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdsToRemove []string) error {
+
+	// Generate the ceph config for running ceph commands similar to the operator
+	if err := writeCephConfig(context, clusterInfo); err != nil {
+		return errors.Wrap(err, "failed to write the ceph config")
+	}
+
+	// Check if the OSD is down
+	osdDump, err := client.GetOSDDump(context, clusterInfo)
+	if err != nil {
+		return errors.Wrap(err, "failed to get osd dump")
+	}
+
+	for _, osdStatus := range osdDump.OSDs {
+		id64, err := osdStatus.OSD.Int64()
+		if err != nil {
+			continue
+		}
+		id := int(id64)
+
+		logger.Debugf("validating status of osd.%d", id)
+
+		status, _, err := osdDump.StatusByID(int64(id))
+		if err != nil {
+			return err
+		}
+
+		const upStatus int64 = 1
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy.It cannot be removed unless it is 'down'", id)
+			continue
+		}
+		logger.Debugf("osd.%d is marked 'DOWN'", id)
+	}
+
+	for _, osdIDStr := range osdsToRemove {
+		logger.Infof("removing OSD %q", osdIDStr)
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+
+		if err := removeOSD(context, clusterInfo, osdID); err != nil {
+			logger.Errorf("failed to remove OSD %d. %v", osdID, err)
+		}
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
+
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+	// Mark the OSD as out. If it is not down, it is expected this command will fail, in which case
+	// we expect to abort purging the OSD
+	output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
+	if err != nil {
+		return errors.Wrapf(err, fmt.Sprintf("cannot remove the osd. %s", output))
+	}
+
+	// Remove the OSD deployment
+	deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+	deployment, err := context.Clientset.AppsV1().Deployments(clusterInfo.Namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, fmt.Sprintf("error fetching the deployment %s", deploymentName))
+	}
+	logger.Infof("removing the OSD deployment %q", deploymentName)
+	if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+		if err != nil {
+			// Continue purging the OSD even if the deployment fails to be deleted
+			logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+		}
+	}
+	if pvcName, ok := deployment.GetLabels()["ceph.rook.io/pvc"]; ok {
+		prepareJobList, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).List(metav1.ListOptions{LabelSelector: "ceph.rook.io/pvc=" + pvcName})
```

Comment:
The osd prepare job has the same label as the osd daemon? I didn't realize that was already there.

Reply:
Yes. osd prepare job and osd deployment/pod have same label.